### ISSUE TITLE
libcontainerd: consider to use task.Wait to update the container's exit code instead of task.Event

### ISSUE THREAD

**Issue by fuweid** [moby/moby#46212]
### Description

Within containerd's current design, if the shim is killed before task-service.Delete API call, the callback on connect close will send 137 exit code because the callback doesn't have any context about container's exit code.

https://github.com/containerd/containerd/blob/70a2c95ae8c02d7a4e448f0e4fb8bb0e6344b5c7/runtime/v2/shim.go#L170-L184

And the moby/moby can't handle duplicate exit event well. Let's say that the moby receives exit code 0 at first and then the duplicate exit event with different code 137 can override 0, as the https://github.com/containerd/containerd/issues/4769 described.

And the containerd task events are best-effort mode because there is no cache and it's handled by shim. I think it's hard to guarantee the exit status is correct. So, I suggest to use task.Wait API instead of event.

REF: https://github.com/containerd/containerd/pull/8954

**Comment by neersighted** [comment]
We discussed this in last week's maintainer's call: overall consensus was that keeping events as the source of truth over using `task.Wait` is preferred; however, we **should** be able to handle duplicate events.

In today's containerd maintainer's call, the fact that we often re-use task IDs (and thus a restarted container might get an event intended for the predecessor under heavy load is possible) was brought up, however, given the presence of the `ExitedAt` field in the protobuf, it should be possible for us to filter events that are intended for a container created before us.

In short: we think that containerd should do what makes the most sense to fix this class of issues robustly, and we think we can adapt Moby to the change if we have to give up on having a reliable one-shot exit event.

**Comment by neersighted** [comment]
We discussed this in last week's maintainers' call: overall consensus was that keeping events as the source of truth over using `task.Wait` is preferred; however, we **should** be able to handle duplicate events.

In today's containerd maintainers' call, the fact that we often re-use task IDs (and thus a restarted container might get an event intended for the predecessor under heavy load is possible) was brought up, however, given the presence of the `ExitedAt` field in the protobuf, it should be possible for us to filter events that are intended for a container created before us.

In short: we think that containerd should do what makes the most sense to fix this class of issues robustly, and we think we can adapt Moby to the change if we have to give up on having a reliable one-shot exit event.

cc @dmcgowan @cpuguy83 @thaJeztah

**Comment by dmcgowan** [comment]
We should address the duplicate event handling on the Moby side before fixing on the containerd side. Any fix to this issue on the containerd side will open up the possibility of duplicate events. If containerd moves the event publishing from shim to containerd, there is still no way to ensure that any shim is not publishing that event.

**Comment by fuweid** [comment]
Closed by #51925 . I will continue to work on containerd side.
